### PR TITLE
fix(drand): `StateGetBeaconEntry` uses chain beacons for historical epochs

### PR DESCRIPTION
**Pull request by rvagg**
Fixes: https://github.com/filecoin-project/lotus/issues/12414

Previously StateGetBeaconEntry would always try and use a drand beacon to get the appropriate round. But as drand has shut down old beacons and we've removed client details from Lotus, it has stopped working for historical beacons.
This fix restores historical beacon entries by using the on-chain lookup, however it now follows the rules used by StateGetRandomnessFromBeacon and the get_beacon_randomness syscall which has some quirks with null rounds prior to nv14. See https://github.com/filecoin-project/lotus/issues/12414#issuecomment-2320034935 for specifics.

StateGetBeaconEntry still blocks for future epochs and uses live drand beacon clients to wait for and fetch rounds as they are available.

---

Notes for review: the existing logic in `chain/rand/rand.go` _should not change_ but I've had to do some refactoring to be able to return the beacon entries as well and not just the randomness. So the file looks a bit different. Please pay special attention to that because it's used for the `get_beacon_randomness` syscall and the `StateGetRandomnessFromBeacon` API. Aside from that, this should only impact `StateGetBeaconEntry`.

### CODE REVIEW

**Comment by rjan90** [comment]
I added a `release/backport` to this PR, and will spin up a Calibration-node with this patch, to confirm that I can get the beacon that Hubert encountered issues with: https://filecoinproject.slack.com/archives/CPFTWMY7N/p1724405387736799?thread_ts=1724401761.517049&cid=CPFTWMY7N

**Review by AnomalRoil — COMMENTED** [review]
Thanks a lot, this should significantly reduce the quantity of nodes trying to query the old defunct incentinet network we lost 👍🏻

File: chain/rand/rand.go
Comment (rvagg) [reply chain/rand/rand.go:168]: yes, it is weird, and I was tempted to rip out the loop and return an error if it's not found in the next non-null epoch. It came in https://github.com/filecoin-project/lotus/pull/7376 and I suspect it was just a rough copy of the logic in `GetLatestBeaconEntry` which searches 20 tipsets for one with a beacon in it and fails after it goes through 20 without finding a single beacon.

With the current logic, even that search should fail, but historically maybe that was not unreasonable?

File: node/impl/full/state_test.go
Comment (rvagg) [reply node/impl/full/state_test.go:95]: no, that's the 20 iterations thing that needs cleaning up:

1. call `getBeaconEntryV3` with the filecoin epoch we want
2. call `GetBeaconRandomnessTipset` on that epoch with `lookback` set to `false`, so it'll skip ahead to the next non-null tipset
3. look in the beacon entries in that tipset for the precise drand round number that we want
4. if we don't find it, walk back up the chain ... but we're not going to find it up the chain because we're matching the exact drand round number so it'll always end in an error, just after a 20 tipset walkback

**Comment by rvagg** [comment]
Opened https://github.com/filecoin-project/lotus/issues/12452 to capture some thoughts about the 20 tipset walkback(s) in the code that I think are unnecessary. I just don't want to adjust the current get-beacon-from-chain logic in this PR so I left it alone.

**Comment by rvagg** [comment]
Calling for objections before I land this. It would be great to get some more review eyes but I understand it's niche.

**Comment by BigLep** [comment]
Per 2024-09-17 FilOz meeting, @jennijuju asked that @Stebalien or @Kubuxu take this to get this landed.

@Stebalien @Kubuxu : see @rvagg 's offer for a sync discussion in https://filecoinproject.slack.com/archives/CP50PPW2X/p1726550218788619 if that is helpful

Also, this one is needed for https://github.com/filecoin-project/lotus/issues/12464 which we want to start this week.

**Comment by BigLep** [comment]
Notes from 2024-09-18 Lotus standup conversation: even though this has been opened for a couple of weeks, it affects pre-quicknet, and it was desired to get in as part of Lotus v1.29.2 this week, we won't merge it until it gets reviewed properly.  We know @Stebalien and @Kubuxu are more tied up this week with a colo, but we assume (and we'll followup) that they'll engage next week.

**Comment by Kubuxu** [comment]
I will try to fetch in the context. I looked at it few weeks ago but I could not reach a conclusion at that time.

File: chain/rand/rand.go
Comment (Kubuxu) [reply chain/rand/rand.go:169]: This logic is for testnets/devnets, which are lower block time networks. In the mainnet, we expect every block to carry a beacon.

File: chain/rand/rand.go
Comment (Kubuxu) [inline chain/rand/rand.go:196]: The `filecoinEpoch` checks change the behaviour here. I'm not sure what the result of `networkVersionGetter(-1)`  is. I'm assuming that it is the network.Version0/1, which would call `getBeaconRandomnessV1`, but it should resolve to genesis randomness either way.

Comment (Kubuxu) [reply chain/rand/rand.go:196]: They also seem redundant/wierd.
IMO: If we want to special case the `filecoinEpoch < 0`, place it at the top of the function before getting networkVersion and explain in a comment.

**Review by Kubuxu — APPROVED** [review]
SGTM, apart from that one comment, where I feel iffy about the intent.

In the future, I suggest separating the commits into non-functional change refactors/renames and functional changes.

File: chain/rand/rand.go
Comment (rvagg) [reply chain/rand/rand.go:169]: oh, this is excellent context, so I've added it as a comment in the two places where we walk back 20 tipsets to find beacon entries; it will mystify future maintainers a little less

File: chain/rand/rand.go
Comment (rvagg) [reply chain/rand/rand.go:203]: This tangle is to deal with the odd `<0` check at the top of the current `V3` method, but I think you're right that my replacement logic is iffy.

I've restored the original branch logic to the top-level function and put the `<0` check back into the `V3`. So it should be closer to whatever the original intent of the `<0` check is. I wish I were able to leave a comment about that though because it's very odd.

@Kubuxu would you mind checking out the latest commit here?

File: chain/store/store.go
Comment (rvagg) [inline chain/store/store.go]: fixed typo here with a second `fixup!` on this branch

**Review by Kubuxu — APPROVED** [review]

**Pull request merged**
